src/biom_subsampled.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * BSD 3-Clause License
+ *
+ * Copyright (c) 2023-2023, UniFrac development team.
+ * All rights reserved.
+ *
+ * See LICENSE file for more details
+ */
+
+#include <cstdlib>
+#include <iostream>
+#include <stdio.h>
+#include <random>
+
+#include "biom_subsampled.hpp"
+
+using namespace su;
+
+linked_sparse_transposed::linked_sparse_transposed(sparse_data &other) 
+  : n_obs(other.n_samples)
+  , n_samples(other.n_obs) {
+    obs_counts_resident = (unsigned int*)calloc(sizeof(unsigned int), n_obs);
+   if(obs_counts_resident == NULL) {
+        fprintf(stderr, "Failed to allocate %zd bytes; [%s]:%d\n", 
+                sizeof(unsigned int) * n_obs, __FILE__, __LINE__);
+        exit(EXIT_FAILURE);
+   }
+
+   for (uint32_t i=0; i<other.n_obs; i++) {
+     const uint32_t cnt = other.obs_counts_resident[i];
+     const uint32_t * idxs = other.obs_indices_resident[i];
+     for (uint32_t j=0; j<cnt; j++) {
+       obs_counts_resident[idxs[j]]++;
+     }
+   }
+
+   obs_data_resident = (double***)malloc(sizeof(double**) * n_obs);
+   if(obs_data_resident == NULL) {
+        fprintf(stderr, "Failed to allocate %zd bytes; [%s]:%d\n", 
+                sizeof(double**) * n_obs, __FILE__, __LINE__);
+        exit(EXIT_FAILURE);
+   }
+   uint32_t _max_count = 0;
+   for (uint32_t i=0; i<n_obs; i++) {
+     const uint32_t cnt = obs_counts_resident[i];
+     _max_count = std::max(_max_count,cnt);
+     obs_data_resident[i] = (double**)malloc(sizeof(double*) * cnt);
+     if(obs_data_resident[i] == NULL) {
+        fprintf(stderr, "Failed to allocate %zd bytes; [%s]:%d\n", 
+                sizeof(double*) * cnt, __FILE__, __LINE__);
+        exit(EXIT_FAILURE);
+     }
+   }
+   max_count = _max_count;
+
+   for (uint32_t i=0; i<other.n_obs; i++) {
+     const uint32_t cnt = other.obs_counts_resident[i];
+     const uint32_t *idxs = other.obs_indices_resident[i];
+     double *data = other.obs_data_resident[i];
+     for (uint32_t j=0; j<cnt; j++) {
+       uint32_t local_i = idxs[j];
+       obs_data_resident[local_i][0] = &(data[j]);
+       obs_data_resident[local_i]++;
+     }
+   }
+
+   // we moved the pointers in place, rewind
+   for (uint32_t i=0; i<n_obs; i++) {
+      const uint32_t cnt = obs_counts_resident[i];
+      obs_data_resident[i] -=  cnt;
+   }
+}
+
+linked_sparse_transposed::~linked_sparse_transposed() {
+    if(obs_data_resident != NULL) {
+        free(obs_data_resident);
+        obs_data_resident = NULL;
+    }
+    if(obs_counts_resident != NULL) {
+       free(obs_counts_resident);
+        obs_counts_resident = NULL;
+    }
+}
+
+void linked_sparse_transposed::transposed_subsample_with_replacement(const uint32_t n, const uint32_t random_seed) {
```

Comment:
In looking close at this, I realized there is an added constraint when using "with replacement": it is necessary to filter out any sample whose counts are below the sampling depth. I realize this _is not_ done by default in the `biom.Table.subsample` method but it arguably should be. The current work around is that we filter the table (i.e., `biom.Table.filter`) in advance of a with replacement call, and that is still viable to do here. What I recommend we do is get this merged first, then consider how to address filtering in a subsequent PR, does that sound good?

Reply:
Yes, this just implements subsampling with replacement as per biom.Table.
If you want a different semantics, we can change it to whatever is the most useful.
We do have the samples_count already, so it would be cheap.

Reply:
@wasade I think I have implemented what you requested.